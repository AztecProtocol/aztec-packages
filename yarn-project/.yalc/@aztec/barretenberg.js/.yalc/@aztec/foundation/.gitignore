dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs