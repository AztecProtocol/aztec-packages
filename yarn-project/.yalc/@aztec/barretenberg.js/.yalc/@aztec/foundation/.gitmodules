[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
