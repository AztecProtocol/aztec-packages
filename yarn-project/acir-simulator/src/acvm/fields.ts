import { AztecAddress, EthAddress, Fr } from '@aztec/foundation';

export type ACVMField = `0x${string}`;
export type ACVMWitness = Map<number, ACVMField>;

export const ZERO_ACVM_FIELD: ACVMField = `0x${Buffer.alloc(32).toString('hex')}`;
export const ONE_ACVM_FIELD: ACVMField = `0x${'00'.repeat(31)}01`;

function adaptBufferSize(originalBuf: Buffer) {
  const buffer = Buffer.alloc(32);
  if (originalBuf.length > buffer.length) {
    throw new Error('Buffer does not fit in 32 bytes');
  }
  originalBuf.copy(buffer, buffer.length - originalBuf.length);
  return buffer;
}

export function toACVMField(value: AztecAddress | EthAddress | Fr | Buffer | boolean | number): `0x${string}` {
  if (typeof value === 'boolean') {
    return value ? ONE_ACVM_FIELD : ZERO_ACVM_FIELD;
  }

  let buffer;

  if (Buffer.isBuffer(value)) {
    buffer = value;
  } else if (typeof value === 'number') {
    buffer = Buffer.alloc(32);
    buffer.writeUInt32BE(value, 28);
  } else {
    buffer = value.toBuffer();
  }

  return `0x${adaptBufferSize(buffer).toString('hex')}`;
}

export function fromACVMField(field: `0x${string}`): Fr {
  const buffer = Buffer.from(field.slice(2), 'hex');
  return Fr.fromBuffer(buffer);
}
