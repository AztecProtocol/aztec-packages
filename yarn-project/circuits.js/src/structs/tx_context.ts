import { BufferReader } from '@aztec/foundation/serialize';

import { FieldsOf, PublicKey } from '../index.js';
import { serializeToBuffer } from '../utils/serialize.js';
import { AztecAddress, EthAddress, Fr, Point } from './index.js';

/**
 * Contract deployment data in a TxContext
 * cpp/src/aztec3/circuits/abis/contract_deployment_data.hpp.
 *
 * Not to be confused with NewContractData.
 */
export class ContractDeploymentData {
  /** Ethereum address of the portal contract on L1. */
  public portalContractAddress: EthAddress;

  constructor(
    /** Public key of the contract deployer (used when deploying account contracts). */
    public deployerPublicKey: PublicKey,
    /** Hash of the constructor verification key. */
    public constructorVkHash: Fr,
    /** Function tree root. */
    public functionTreeRoot: Fr,
    /** Contract address salt (used when deriving a contract address). */
    public contractAddressSalt: Fr,
    /**
     * Ethereum address of the portal contract on L1.
     * TODO(AD): union type kludge due to cbind compiler having special needs
     */
    portalContractAddress: EthAddress | AztecAddress,
  ) {
    this.portalContractAddress = EthAddress.fromField(portalContractAddress.toField());
  }

  toBuffer() {
    return serializeToBuffer(
      this.deployerPublicKey,
      this.constructorVkHash,
      this.functionTreeRoot,
      this.contractAddressSalt,
      this.portalContractAddress,
    );
  }

  /**
   * Returns an empty ContractDeploymentData.
   * @returns The empty ContractDeploymentData.
   */
  public static empty(): ContractDeploymentData {
    return new ContractDeploymentData(Point.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, EthAddress.ZERO);
  }
  /**
   * Deserializes contract deployment data rom a buffer or reader.
   * @param buffer - Buffer to read from.
   * @returns The deserialized ContractDeploymentData.
   */
  static fromBuffer(buffer: Buffer | BufferReader): ContractDeploymentData {
    const reader = BufferReader.asReader(buffer);
    return new ContractDeploymentData(
      reader.readObject(Point),
      reader.readFr(),
      reader.readFr(),
      reader.readFr(),
      new EthAddress(reader.readBytes(32)),
    );
  }
}

/**
 * Transaction context.
 * @see cpp/src/aztec3/circuits/abis/tx_context.hpp.
 */
export class TxContext {
  constructor(
    /**
     * Whether this is a fee paying tx. If not other tx in a bundle will pay the fee.
     */
    public isFeePaymentTx: boolean,
    /**
     * Indicates whether this a gas rebate payment tx.
     *
     * NOTE: The following is a WIP and it is likely to change in the future.
     * Explanation: Each tx is actually 3 txs in one: a fee-paying tx, the actual tx you want to execute, and a rebate
     * tx. The fee-paying tx pays some `max_fee = gas_price * gas_limit`. Then the actual tx will cost an amount of gas
     * to execute (actual_fee = gas_price * gas_used). Then the rebate tx returns `max_fee - actual_fee` back to
     * the user.
     */
    public isRebatePaymentTx: boolean,
    /**
     * Whether this is a contract deployment tx.
     */
    public isContractDeploymentTx: boolean,
    /**
     * Contract deployment data.
     */
    public contractDeploymentData: ContractDeploymentData,
    /**
     * Chain ID of the transaction. Here for replay protection.
     */
    public chainId: Fr,
    /**
     * Version of the transaction. Here for replay protection.
     */
    public version: Fr,
  ) {}

  /**
   * Serialize as a buffer.
   * @returns The buffer.
   */
  toBuffer() {
    return serializeToBuffer(
      this.isFeePaymentTx,
      this.isRebatePaymentTx,
      this.isContractDeploymentTx,
      this.contractDeploymentData,
      this.chainId,
      this.version,
    );
  }

  /**
   * Deserializes TxContext from a buffer or reader.
   * @param buffer - Buffer to read from.
   * @returns The TxContext.
   */
  static fromBuffer(buffer: Buffer | BufferReader): TxContext {
    const reader = BufferReader.asReader(buffer);
    return new TxContext(
      reader.readBoolean(),
      reader.readBoolean(),
      reader.readBoolean(),
      reader.readObject(ContractDeploymentData),
      reader.readFr(),
      reader.readFr(),
    );
  }

  static empty(chainId: Fr | number = 0, version: Fr | number = 0) {
    return new TxContext(false, false, false, ContractDeploymentData.empty(), new Fr(chainId), new Fr(version));
  }

  /**
   * Create a new instance from a fields dictionary.
   * @param fields - The dictionary.
   * @returns A new instance.
   */
  static from(fields: FieldsOf<TxContext>): TxContext {
    return new TxContext(...TxContext.getFields(fields));
  }

  /**
   * Serialize into a field array. Low-level utility.
   * @param fields - Object with fields.
   * @returns The array.
   */
  static getFields(fields: FieldsOf<TxContext>) {
    return [
      fields.isFeePaymentTx,
      fields.isRebatePaymentTx,
      fields.isContractDeploymentTx,
      fields.contractDeploymentData,
      fields.chainId,
      fields.version,
    ] as const;
  }
}
