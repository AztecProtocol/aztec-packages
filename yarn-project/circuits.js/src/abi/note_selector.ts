import { toBigIntBE } from '@aztec/foundation/bigint-buffer';
import { randomInt } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { hexSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, TypeRegistry } from '@aztec/foundation/serialize';

import { Selector } from './selector.js';

/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */

/** Note selector branding */
export interface NoteSelector {
  /** Brand. */
  _branding: 'NoteSelector';
}

/**
 * A note selector is a 7 bit long value that identifies a note type within a contract.
 * TODO(#10952): Encoding of note type id can be reduced to 7 bits.
 */
export class NoteSelector extends Selector {
  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param buffer - Buffer  or BufferReader to read from.
   * @returns The Selector.
   */
  static fromBuffer(buffer: Buffer | BufferReader) {
    const reader = BufferReader.asReader(buffer);
    const value = Number(toBigIntBE(reader.readBytes(Selector.SIZE)));
    if (value >= 1 << 7) {
      throw new Error(`Invalid note selector: ${value}`);
    }
    return new NoteSelector(value);
  }

  static fromString(buf: string) {
    const withoutPrefix = buf.replace(/^0x/i, '').slice(-8);
    const buffer = Buffer.from(withoutPrefix, 'hex');
    return NoteSelector.fromBuffer(buffer);
  }

  /**
   * Converts a field to selector.
   * @param fr - The field to convert.
   * @returns The selector.
   */
  static fromField(fr: Fr) {
    return new NoteSelector(Number(fr.toBigInt()));
  }

  /**
   * Creates an empty selector.
   * @returns An empty selector.
   */
  static empty() {
    return new NoteSelector(0);
  }

  /**
   * Creates a random selector.
   * @returns A random selector.
   */
  static random() {
    const value = randomInt(1 << 7);
    return NoteSelector.fromField(new Fr(value));
  }

  toJSON() {
    return this.toString();
  }

  static get schema() {
    return hexSchemaFor(NoteSelector);
  }
}

// For deserializing JSON.
TypeRegistry.register('NoteSelector', NoteSelector);
