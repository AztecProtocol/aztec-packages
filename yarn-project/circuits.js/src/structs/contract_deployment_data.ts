import { pedersenHash } from '@aztec/foundation/crypto';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr, Point } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { FieldsOf } from '@aztec/foundation/types';

import { CONTRACT_DEPLOYMENT_DATA_LENGTH, GeneratorIndex } from '../constants.gen.js';
import { PublicKey } from '../types/public_key.js';

/**
 * Contract deployment data in a TxContext
 * Not to be confused with NewContractData.
 */
export class ContractDeploymentData {
  constructor(
    /** Public key of the contract. */
    public publicKey: PublicKey,
    /** Hash of the initialization payload. */
    public initializationHash: Fr,
    /** Contract class identifier. */
    public contractClassId: Fr,
    /** Contract address salt (used when deriving a contract address). */
    public contractAddressSalt: Fr,
    /** Ethereum address of the portal contract on L1. */
    public portalContractAddress: EthAddress,
  ) {}

  static getFields(fields: FieldsOf<ContractDeploymentData>) {
    return [
      fields.publicKey,
      fields.initializationHash,
      fields.contractClassId,
      fields.contractAddressSalt,
      fields.portalContractAddress,
    ] as const;
  }

  toBuffer() {
    return serializeToBuffer(...ContractDeploymentData.getFields(this));
  }

  toFields(): Fr[] {
    const fields = serializeToFields(...ContractDeploymentData.getFields(this));
    if (fields.length !== CONTRACT_DEPLOYMENT_DATA_LENGTH) {
      throw new Error(
        `Invalid number of fields for ContractDeploymentData. Expected ${CONTRACT_DEPLOYMENT_DATA_LENGTH}, got ${fields.length}`,
      );
    }
    return fields;
  }

  /**
   * Returns an empty ContractDeploymentData.
   * @returns The empty ContractDeploymentData.
   */
  public static empty(): ContractDeploymentData {
    return new ContractDeploymentData(Point.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, EthAddress.ZERO);
  }

  isEmpty() {
    return ContractDeploymentData.getFields(this).every(f => f.isZero());
  }

  /**
   * Deserializes contract deployment data rom a buffer or reader.
   * @param buffer - Buffer to read from.
   * @returns The deserialized ContractDeploymentData.
   */
  static fromBuffer(buffer: Buffer | BufferReader): ContractDeploymentData {
    const reader = BufferReader.asReader(buffer);
    return new ContractDeploymentData(
      reader.readObject(Point),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      reader.readObject(EthAddress),
    );
  }

  static fromFields(fields: Fr[] | FieldReader): ContractDeploymentData {
    const reader = FieldReader.asReader(fields);
    return new ContractDeploymentData(
      reader.readObject(Point),
      reader.readField(),
      reader.readField(),
      reader.readField(),
      reader.readObject(EthAddress),
    );
  }

  hash(): Fr {
    return pedersenHash(
      this.toFields().map(f => f.toBuffer()),
      GeneratorIndex.CONTRACT_DEPLOYMENT_DATA,
    );
  }
}
