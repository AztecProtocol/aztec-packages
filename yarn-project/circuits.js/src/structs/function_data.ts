import { FunctionAbiHeader, FunctionSelector, FunctionType } from '@aztec/foundation/abi';
import { BufferReader } from '@aztec/foundation/serialize';

import { ContractFunctionDao } from '../index.js';
import { serializeToBuffer } from '../utils/serialize.js';

/**
 * Function description for circuit.
 * @see abis/function_data.hpp
 */
export class FunctionData {
  constructor(
    /**
     * Function selector of the function being called.
     */
    public selector: FunctionSelector,
    /**
     * Indicates whether the function is only callable by self or not.
     */
    public isInternal: boolean,
    /**
     * Indicates whether the function is private or public.
     */
    public isPrivate: boolean,
    /**
     * Indicates whether the function is a constructor.
     */
    public isConstructor: boolean,
  ) {}

  static fromAbi(abi: FunctionAbiHeader | ContractFunctionDao): FunctionData {
    return new FunctionData(
      FunctionSelector.fromNameAndParameters(abi.name, abi.parameters),
      abi.isInternal,
      abi.functionType === FunctionType.SECRET,
      abi.name === 'constructor',
    );
  }

  /**
   * Serialize this as a buffer.
   * @returns The buffer.
   */
  toBuffer(): Buffer {
    return serializeToBuffer(this.selector, this.isInternal, this.isPrivate, this.isConstructor);
  }

  /**
   * Returns whether this instance is empty.
   * @returns True if the function selector is zero.
   */
  isEmpty() {
    return this.selector.isEmpty();
  }

  /**
   * Returns a new instance of FunctionData with zero function selector.
   * @param args - Arguments to pass to the constructor.
   * @returns A new instance of FunctionData with zero function selector.
   */
  public static empty(args?: {
    /**
     * Indicates whether the function is only callable by self or not.
     */
    isInternal?: boolean;
    /**
     * Indicates whether the function is private or public.
     */
    isPrivate?: boolean;
    /**
     * Indicates whether the function is a constructor.
     */
    isConstructor?: boolean;
  }): FunctionData {
    return new FunctionData(
      FunctionSelector.empty(),
      args?.isInternal ?? false,
      args?.isPrivate ?? false,
      args?.isConstructor ?? false,
    );
  }

  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param buffer - Buffer or reader to read from.
   * @returns A new instance of FunctionData.
   */
  static fromBuffer(buffer: Buffer | BufferReader): FunctionData {
    const reader = BufferReader.asReader(buffer);
    return new FunctionData(
      reader.readObject(FunctionSelector),
      reader.readBoolean(),
      reader.readBoolean(),
      reader.readBoolean(),
    );
  }
}
