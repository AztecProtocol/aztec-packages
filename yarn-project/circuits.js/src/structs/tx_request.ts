import { AztecAddress } from '@aztec/foundation/aztec-address';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';

import { FieldsOf } from '../utils/jsUtils.js';
import { serializeToBuffer } from '../utils/serialize.js';
import { FunctionData } from './function_data.js';
import { TxContext } from './tx_context.js';

/**
 * Transaction request.
 * @see cpp/src/aztec3/circuits/abis/tx_request.hpp.
 */
export class TxRequest {
  constructor(
    /**
     * Sender.
     */
    public origin: AztecAddress,
    /**
     * Function data representing the function to call.
     */
    public functionData: FunctionData,
    /**
     * Pedersen hash of function arguments.
     */
    public argsHash: Fr,
    /**
     * Transaction context.
     */
    public txContext: TxContext,
  ) {}

  static getFields(fields: FieldsOf<TxRequest>) {
    return [fields.origin, fields.functionData, fields.argsHash, fields.txContext] as const;
  }

  static from(fields: FieldsOf<TxRequest>): TxRequest {
    return new TxRequest(...TxRequest.getFields(fields));
  }

  /**
   * Serialize as a buffer.
   * @returns The buffer.
   */
  toBuffer() {
    const fields = TxRequest.getFields(this);
    return serializeToBuffer([...fields]);
  }

  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param buffer - Buffer to read from.
   * @returns The deserialised TxRequest object.
   */
  static fromBuffer(buffer: Buffer | BufferReader): TxRequest {
    const reader = BufferReader.asReader(buffer);
    return new TxRequest(
      reader.readObject(AztecAddress),
      reader.readObject(FunctionData),
      reader.readFr(),
      reader.readObject(TxContext),
    );
  }
}
