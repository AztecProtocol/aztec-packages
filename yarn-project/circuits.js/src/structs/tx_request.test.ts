import { FunctionSelector } from '@aztec/foundation/abi';
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { Fr, Point } from '@aztec/foundation/fields';

import { TX_REQUEST_LENGTH } from '../constants.gen.js';
import { makeTxRequest } from '../tests/factories.js';
import { ContractDeploymentData } from './contract_deployment_data.js';
import { FunctionData } from './function_data.js';
import { EthAddress } from './index.js';
import { TxContext } from './tx_context.js';
import { TxRequest } from './tx_request.js';

describe('TxRequest', () => {
  let request: TxRequest;

  beforeAll(() => {
    const randomInt = Math.floor(Math.random() * 1000);
    request = makeTxRequest(randomInt);
  });

  it(`serializes to buffer and deserializes it back`, () => {
    const buffer = request.toBuffer();
    const res = TxRequest.fromBuffer(buffer);
    expect(res).toEqual(request);
    expect(res.isEmpty()).toBe(false);
  });

  it('number of fields matches constant', () => {
    const fields = request.toFields();
    expect(fields.length).toBe(TX_REQUEST_LENGTH);
  });

  it('compute hash', () => {
    const deploymentData = new ContractDeploymentData(
      new Point(new Fr(1), new Fr(2)),
      new Fr(1),
      new Fr(2),
      new Fr(3),
      EthAddress.fromField(new Fr(1)),
    );
    const txRequest = TxRequest.from({
      origin: AztecAddress.fromBigInt(1n),
      functionData: new FunctionData(FunctionSelector.fromField(new Fr(2n)), false, true, true),
      argsHash: new Fr(3),
      txContext: new TxContext(false, false, true, deploymentData, Fr.ZERO, Fr.ZERO),
    });

    const hash = txRequest.hash().toString();

    expect(hash).toMatchSnapshot();

    // Value used in hash test in tx_request.nr
    // console.log("hash", hash);
  });
});
