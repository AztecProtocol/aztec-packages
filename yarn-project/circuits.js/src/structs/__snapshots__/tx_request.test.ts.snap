// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TxRequest compute hash 1`] = `"0x05e3f67a5e787faf63c85c28169a9d616b594c9799d9785b4f167d9cd7a71e9e"`;
